src/webgpu/util/fp_interval.ts

```diff
@@ -0,0 +1,238 @@
+    } else {
+      // x_32 rounded towards -inf, so is before x
+      const otherside = nextAfter(x_32, false, false).value as number;
+      return new FPInterval(converted, otherside);
+    }
+  }
+}
+
+class ULPFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  protected readonly n: number;
+
+  public constructor(n: number) {
+    this.n = Math.abs(n);
+  }
+
+  public singular(x: number): FPInterval {
+    assert(!Number.isNaN(x), `ULP not defined for NaN`);
+    return FPInterval.span(this.impl(x), this.impl(flushSubnormalNumber(x)));
+  }
+
+  public range(i: FPInterval): FPInterval {
+    if (i.begin === i.end) {
+      return this.singular(i.begin);
+    }
+    return FPInterval.span(this.singular(i.begin), this.singular(i.end));
+  }
+
+  private impl(x: number): FPInterval {
+    const ulp_flush = oneULP(x, true);
+    const ulp_noflush = oneULP(x, false);
+    const ulp = Math.max(ulp_flush, ulp_noflush);
+
+    return new FPInterval(x - this.n * ulp, x + this.n * ulp);
+  }
+}
+
+export class CosFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  private readonly builder;
+
+  public constructor() {
+    this.builder = new AbsoluteFPIntervalBuilder(2 ** -11);
+  }
+
+  public singular(x: number): FPInterval {
+    return FPInterval.span(this.impl(x), this.impl(flushSubnormalNumber(x)));
+  }
+
+  private impl(x: number): FPInterval {
+    return this.builder.singular(Math.cos(x));
+  }
+}
+
+export class SinFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  private readonly builder;
+
+  public constructor() {
+    this.builder = new AbsoluteFPIntervalBuilder(2 ** -11);
+  }
+
+  public singular(x: number): FPInterval {
+    return FPInterval.span(this.impl(x), this.impl(flushSubnormalNumber(x)));
+  }
+
+  private impl(x: number): FPInterval {
+    return this.builder.singular(Math.sin(x));
+  }
+}
+
+export class DivisionFPIntervalBuilder implements FPBinaryIntervalBuilder {
+  private readonly builder;
+
+  public constructor() {
+    this.builder = new ULPFPIntervalBuilder(2.5);
+  }
+
+  public singular(x: number, y: number): FPInterval {
+    assert(y !== 0, `division by 0 is not defined`);
+    return FPInterval.span(this.impl(x, y), this.impl(flushSubnormalNumber(x), y));
+  }
+
+  public range(ix: FPInterval, iy: FPInterval): FPInterval {
+    if (ix.begin === ix.end && iy.begin === iy.end) {
+      return this.singular(ix.begin, iy.begin);
+    }
+    return FPInterval.span(
+      this.singular(ix.begin, iy.begin),
+      this.singular(ix.begin, iy.end),
+      this.singular(ix.end, iy.begin),
+      this.singular(ix.end, iy.end)
+    );
+  }
+
+  private impl(x: number, y: number): FPInterval {
+    return this.builder.singular(x / y);
+  }
+}
+
+export class TanFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  private readonly cos_builder;
+  private readonly sin_builder;
+  private readonly divide_builder;
+
+  public constructor() {
+    this.cos_builder = new CosFPIntervalBuilder();
```

Comment:
In a production version, since these builders are always the same, I will probably figure out a way to properly cache them or make a table of them, instead of having each one creating instances. Possibly just use static classes/methods